### PR TITLE
Mostly revert the filter code to reduce its complexity of use.

### LINKED ISSUE #1836: Take nc_filter_actions() out of dispatch table, and add nc_def_var_filter() and nc_inq_var_filter() instead, keeping nc_filter_actions() internal to netcdf-c
The filter_actions function in the dispatch table requires some constants from ncfilter.h, which is not installed.

So I will have to copy these constants to PIO code in order to have them, which is not ideal. Either ncfilter.h needs to be installed, or else these constants (and perhaps everything in ncfilter.h) needs to be moved to netcdf_filter.h, which is installed.

The constants are:
```
/* Internal filter actions */
#define NCFILTER_DEF		1
#define NCFILTER_REMOVE  	2
#define NCFILTER_FILTERIDS      3
#define NCFILTER_INFO		4
#define NCFILTER_CLIENT_REG	10
#define NCFILTER_CLIENT_UNREG	11
#define NCFILTER_CLIENT_INQ	12

```

### PR DESCRIPTION
**Pull request by DennisHeimbigner**
* Fixes #1836 

re: https://github.com/Unidata/netcdf-c/issues/1836

Revert the internal filter code to simplify it. From the user's
point of view, the only visible changes should be:

1. The functions that convert text to filter specs have had their signature reverted and have been moved to netcdf_aux.h
2. Some filter API functions now return NC_ENOFILTER when inquiry is made about some filter.

Internally,the dispatch table has been modified to get rid of the filter_actions
entry and associated complex structures. It has been replaced with
inq_var_filter_ids and inq_var_filter_info entries and the dispatch table
version has been bumped to 3. Corresponding NOOP and NOTNC4 functions
were added to libdispatch/dnotnc4.c. Also, the filter_action entries
in dispatch tables were replaced for all dispatch code bases (HDF5, DAP2,
etc). This should only impact UDF users.

In the process, it became clear that the form of the filters
field in NC_VAR_INFO_T was format dependent, so I converted it to
be of type void* and pushed its management into the various dispatch
code bases. Specifically libhdf5 and libnczarr now manage the filters
field in their own way.

The auxilliary functions for parsing textual filter specifications
were moved to netcdf_aux.h and were renamed to the following:
* ncaux_h5filterspec_parse
* ncaux_h5filterspec_parselist
* ncaux_h5filterspec_free
* ncaux_h5filter_fix8

Misc. Other Changes:

1. Document NUG/filters.md updated to reflect the changes above.
2. All the old data types (structs and enums)
   used by filter_actions actions were deleted.
   The exception is the NC_H5_Filterspec because it is needed
   by ncaux_h5filterspec_parselist.
3. Clientside filters were removed -- another enhancement
   for which no-one ever asked.
4. The ability to remove filters was itself removed.
5. Some functionality needed by nczarr was moved from libhdf5
   to libsrc4 e.g. nc4_find_default_chunksizes
6. All the filterx code was removed
7. ncfilter.h and nc4filter.c no longer used

Misc. Unrelated Changes:

1. The nczarr_test makefile clean was leaving some directories; so
   add clean-local to take care of them.

### CODE REVIEW

File: libsrc4/nc4var.c [hunk 1]
```diff
@@ -1463,3 +1466,169 @@ nc4_get_typelen_mem(NC_FILE_INFO_T *h5, nc_type xtype, size_t *len)
 
     return NC_NOERR;
 }
+
+
+/**
+ * @internal Check a set of chunksizes to see if they specify a chunk
```
Comment (captainkirk99) [inline libsrc4/nc4var.c:1472]: Are you moving the chunksize code because chunks now occur in multiple dispatch layers (HDF5 and zarr)?

**Comment by DennisHeimbigner** [comment]
Yes, I am moving (as needed) stuff in libhdf5 to libsrc4 if it is being used
in libnczarr as well. In this case, it was an indirect consequence of the filter
code modifications because setting chunksizes is a potential by product
of adding a filter. There are almost certainly other code moves that will
need to occur from libhdf5 to libsrc4. In some cases, I may have to refactor
the code because sometimes there is an hdf5 dependency deep in the code
so I cannot just move the code wholesale, but need to break it into smaller
pieces that are hdf5 independent.

**Comment by WardF** [comment]
Thanks Dennis, reviewing now!

**Comment by edwardhartnett** [comment]
@DennisHeimbigner if you need help with the libhdf5/libsrc4 code let me know.

**Comment by DennisHeimbigner** [comment]
Thanks Ed. This is uncharted territory since zarr is the first netcdf-4 but not hdf5
code base we have had.

**Comment by edwardhartnett** [comment]
Well the HDF4 code also uses libsrc4, but since it's a read-only interface, it doesn't use as much of it.

**Comment by WardF** [comment]
So this looks good to go from a technical standpoint; thanks Dennis.  I've reviewed the ABI comparison documentation (attached) and I'm a little concerned that this is going to break people's code downstream, but I'm not certain what to do about it at this point.  I say this because of the removed public API functions.  At the very least maybe we should have a short blog post we can reference people to which outlines the logic/justification behind this decision?  What do you think, @DennisHeimbigner ?

* ABI Report: [revert-filters-abi.zip](https://github.com/Unidata/netcdf-c/files/5313620/revert-filters-abi.zip)

**Comment by DennisHeimbigner** [comment]
There is a section near the front of NUG/filters.md that attempts
to capture these incompatibilities.  It appears to cover the issues
in the ABI compatibility report. As noted in the filters.md (Appendix B),
it is possible to detect these using a existence of NC_ENOFILTER and/or
NC_HAS_MULTIFILTER.
We could turn this into a blog post if you would like.

**Comment by DennisHeimbigner** [comment]
I made a slight change to the incompatibilities section to link it to
appendix B and vice-versa.

**Comment by WardF** [comment]
I think getting in front of this would be better so that people are aware its coming; the documentation is great, but I'm thinking more of getting the info to our community prior to the upcoming Zarr-enabled pre-release, rather than pointing them to documentation they'll only see/seek out once their code is broken.

**Comment by DennisHeimbigner** [comment]
Here is the proposed blog entry to be posted after a new release has been made.
````
Netcdf-4 Filter Support Changes

The netcdf-c library filters API in version 4.7.4 has been
deprecated in favor of a modified version that unfortunately
may cause incompatibilties for users.

The initial reason for the incompatible changes was to support
the use of filters in the new NCZarr code. The changes were not
completely thought out so it was decided to remove them and
revert to previous behaviors. At some future point, the filter
mechanism will be extended to support filters for NCZarr, but
these will be proper extensions: the existing, reverted, filter
API will continue to be supported with no user-visible
modifications.

Unfortunately, some advanced users of netcdf filters may
experience some compilation or execution problems for previously
working code because of these reversions. In that case, please
revise your code. Apologies are extended for any inconvenience.
Note that it is possible to detect which mechanism is in place.
See the netcdf-c github document NUG/filters.md for details.
````

**Comment by edwardhartnett** [comment]
Even the great Homer nods...

**Comment by WardF** [comment]
> Here is the proposed blog entry to be posted after a new release has been made.
> 
> ```
> Netcdf-4 Filter Support Changes
> 
> The netcdf-c library filters API in version 4.7.4 has been
> deprecated in favor of a modified version that unfortunately
> may cause incompatibilties for users.
> 
> The initial reason for the incompatible changes was to support
> the use of filters in the new NCZarr code. The changes were not
> completely thought out so it was decided to remove them and
> revert to previous behaviors. At some future point, the filter
> mechanism will be extended to support filters for NCZarr, but
> these will be proper extensions: the existing, reverted, filter
> API will continue to be supported with no user-visible
> modifications.
> 
> Unfortunately, some advanced users of netcdf filters may
> experience some compilation or execution problems for previously
> working code because of these reversions. In that case, please
> revise your code. Apologies are extended for any inconvenience.
> Note that it is possible to detect which mechanism is in place.
> See the netcdf-c github document NUG/filters.md for details.
> ```

Looks good! I'd say add a link to the blog post in the release notes (or we can remember to do it after the fact) and I'll get this merged in and we can move towards the ncZarr pre-release we've been asked about.

**Pull request merged**